### ISSUE TITLE
Stop training earlier

### ISSUE THREAD

**Issue by mdeeds** [mdeeds/genie#4]
We should probably stop training the model as soon as it gets 100% on all training data. Don't continue to push labels all the way down to 1 or 0. Instead, terminate when all classifications on the training set are correct and outside the 0.3 to 0.7 range.

We might want to implement a custom loss function where there is no benefit of pushing a label past ~0.9 or ~0.1.  Maybe this is not necessary since we use a hard sigmoid, so as soon as the penultimate layer has pushed it to 1/0, there is no benefit to push further.

Judging from this CodePen: https://codepen.io/caisq/pen/QZYZEZ it seems that a custom loss function is just a function which takes two tensors and returns a tensor with the loss.

```
const pred = tf.tensor1d([0.1, 0.2, 0.5, 0.8, 0.9]);
const actual = tf.tensor1d([1, 1, 1, 1, 1]);
const absDifference = tf.abs(tf.sub(pred, actual));
const floor = tf.tensor([0.3], []);
const error = tf.maximum(floor, absDifference);
error.print()

Tensor
    [0.9, 0.8, 0.5, 0.3, 0.3]
```

**Comment by SteveDeeds** [comment]
It looks like the learning rate has very little to do with how many samples are needed to get all the predictions correct:
![image](https://user-images.githubusercontent.com/70077400/112688103-a0602b00-8e35-11eb-996c-a9371b34648e.png)

**Comment by SteveDeeds** [comment]
Learning rate of 0.001 seems to require the least number of examples for 500 epochs.  Although, there is not too much of a difference between the number of examples required at the different learning rates. 
![image](https://user-images.githubusercontent.com/70077400/112700571-b4168c00-8e4b-11eb-8b0d-c91d5374e245.png)

**Comment by SteveDeeds** [comment]
A learning rate of 1 seems to produce good results quickly.  This is a much higher rate than I expected.  Perhaps with more complicated games a lower learning rate would be required.
![image](https://user-images.githubusercontent.com/70077400/112707248-33658900-8e67-11eb-8e4b-31a12a4ce344.png)

**Comment by mdeeds** [comment]
Amazing.  This is not what I expected.  The relationship is very linear on
the left side (on the log-log scale).  This seems to indicate the learning
rates less than 0.1 are not helpful - it's going in the right direction,
just taking smaller steps. Around 0.08 or so, it seems to get better with
higher rates at a super-linear rate.  I guess what happens here is that the
gradient is in the "right" direction with 100 examples, and because we use
a hard-sigmoid, as long as we push it past that hard edge, it gets a
perfect score.  I would think that if you decreased the number of examples
to ~5 or fewer, we'd see the opposite effect where the higher learning
rates do not work as well.

**Comment by SteveDeeds** [comment]
Interesting.  I made some optimizations based on this data and pushed a change that stops training early, so I'm going to close this.
